File: bindings/python/examples/fingerprint.py
"""fingerprint — prove the client really impersonates Chrome by reading back the
JA4 / JA3 / Akamai fingerprints a server computes from our TLS + HTTP/2 bytes,
then confirming the H3 path actually negotiates over QUIC.

``tls.peet.ws`` echoes the fingerprints it observed over HTTP/2 (it is a TCP/H2
oracle and does not serve HTTP/3). For H3 we just connect to a known QUIC host
and check the negotiated ALPN is ``h3``.

    python examples/fingerprint.py
"""
import holytls


def main():
    # HTTP/2: read the fingerprints the oracle computed from our bytes.
    with holytls.Client(timeout_ms=30000) as client:
        r = client.get("https://tls.peet.ws/api/all")
        r.raise_for_status()
        j = r.json()
        tls = j.get("tls", {})
        print(f"HTTP/2 (ALPN {r.alpn}):")
        print("  JA4      :", tls.get("ja4"))
        print("  JA3 hash :", tls.get("ja3_hash"))
        print("  peetprint:", tls.get("peetprint_hash"))
        print("  Akamai H2:", (j.get("http2") or {}).get("akamai_fingerprint_hash"))

    # HTTP/3: confirm the QUIC path negotiates h3 against a real H3 host.
    with holytls.Client(http_version="h3", timeout_ms=30000) as client:
        r = client.get("https://cloudflare-quic.com/")
        print(f"\nHTTP/3: ok={r.ok} status={r.status_code} alpn={r.alpn!r}")


if __name__ == "__main__":
    main()

File: examples/fingerprint.c
// examples/fingerprint.c — prove the impersonation. holytls reproduces a real
// Chrome's TLS + HTTP/2 fingerprint byte-for-byte. This fetches tls.peet.ws —
// an oracle that reports the fingerprint IT observed on the wire — and prints
// the JA3 / JA4 (TLS) and Akamai (HTTP/2) fingerprints the server saw. Point a
// real Chrome 149 at the same URL and the hashes match.
//
// Run:  ./build/fingerprint
#include <stdio.h>

#include "base/arena.h"
#include "base/base.h"
#include "base/string8.h"
#include "core/client.h"
#include "core/json.h"
#include "net/loop.h"
#include "profile/profile.h"

typedef struct Ctx Ctx;
struct Ctx {
  B32 done;
};

static void on_response(void *user, const Response *resp) {
  Ctx *cx = (Ctx *)user;
  cx->done = 1;
  if (!resp->ok) {
    fprintf(stderr, "request failed: %s\n", resp->error ? resp->error : "?");
    return;
  }

  // Parse the JSON body into a scratch arena — the doc and the String8 views
  // into it live until we release the arena. response_json(resp, a) is just
  // json_parse(a, response_text(resp)).
  Arena *a = arena_alloc();
  yyjson_doc *doc = response_json(resp, a);
  if (!doc) {
    fprintf(stderr, "could not parse the tls.peet.ws response as JSON\n");
    arena_release(a);
    return;
  }

  // tls.peet.ws/api/all reports what it observed. These are the fingerprints
  // anti-bot systems key on; holytls reproduces Chrome's exactly.
  printf("negotiated ALPN : " STR8_Fmt "\n", STR8_Arg(resp->alpn));
  printf("user-agent      : " STR8_Fmt "\n",
         STR8_Arg(json_ptr_str(doc, "/user_agent")));
  printf("TLS  JA3 hash   : " STR8_Fmt "\n",
         STR8_Arg(json_ptr_str(doc, "/tls/ja3_hash")));
  printf("TLS  JA4        : " STR8_Fmt "\n",
         STR8_Arg(json_ptr_str(doc, "/tls/ja4")));
  printf("TLS  PeetPrint# : " STR8_Fmt "\n",
         STR8_Arg(json_ptr_str(doc, "/tls/peetprint_hash")));
  printf("HTTP/2 Akamai   : " STR8_Fmt "\n",
         STR8_Arg(json_ptr_str(doc, "/http2/akamai_fingerprint")));
  printf("HTTP/2 Akamai#  : " STR8_Fmt "\n",
         STR8_Arg(json_ptr_str(doc, "/http2/akamai_fingerprint_hash")));

  arena_release(a);
}

int main(void) {
  EventLoop loop;
  loop_init(&loop);

  // Chrome 149 over HTTP/2 — the Akamai (HTTP/2) fingerprint needs an H2
  // connection, and JA3/JA4 come from the same ClientHello either way.
  Client client;
  client_init(&client, &loop, profile_chrome149(), /*verify=*/1);
  client_set_timeout_ms(&client, 30000);

  Ctx cx = {0};
  client_get(&client, str8_lit("https://tls.peet.ws/api/all"), on_response,
             &cx);
  loop_run(&loop);

  client_cleanup(&client);
  loop_shutdown(&loop);
  return cx.done ? 0 : 1;
}
